### ISSUE TITLE
Reimplementation of QAT support for compression and checksums with Data Plane and and  Intel QAT1.7.L.4.3.0-00033 drivers

### ISSUE THREAD

**Issue by TheTDD** [zfsonlinux/zfs#8076]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  debian
Distribution Version	|  9.5
Linux Kernel	| 4.18.6-rt3
Architecture	| amd64
ZFS Version	| 0.7.11-1
SPL Version	| 0.7.11-1

zfs can' be compiled with QAT 1.7

```
./configure \
        --with-qat=/opt/QAT
```

```
checking qat source directory... /opt/QAT
checking qat build directory... /opt/QAT/build
configure: error:
                *** Please make sure the qat driver is installed then try again.
                *** Failed to find icp_qa_al.ko or qat_api.ko in:
                /opt/QAT/build

```
```
root@linux:/opt/QAT# lsmod | grep qat
qat_dh895xcc           20480  1
intel_qat             253952  30 usdm_drv,qat_dh895xcc
dh_generic             16384  1 intel_qat
uio                    20480  29 intel_qat
authenc                16384  1 intel_qat
```
```
root@linux:/opt/QAT# /usr/bin/adf_ctl status
Checking status of all devices.
There is 1 QAT acceleration device(s) in the system:
 qat_dev0 - type: dh895xcc,  inst_id: 0,  node_id: 0,  bsf: 67:00.0,  #accel: 6 #engines: 12 state: up
```

```
root@linux:/opt/QAT# modinfo intel_qat
filename:       /lib/modules/4.18.6-rt3-tdd/updates/drivers/crypto/qat/qat_common/intel_qat.ko
version:        4.2.0
alias:          crypto-intel_qat
alias:          intel_qat
description:    Intel(R) QuickAssist Technology
author:         Intel
license:        Dual BSD/GPL
srcversion:     E2FCA0EF07383A249963CB7
depends:        uio,dh_generic,authenc
retpoline:      Y
name:           intel_qat
vermagic:       4.18.6-rt3-tdd SMP mod_unload modversions
```

```
root@linux:/opt/QAT/build# ls -al
total 6768
drwxr-xr-x 2 root root    4096 Oct 31 15:34 .
drwxr-xr-x 4 root root    4096 Oct 31 15:34 ..
-rwxr-x--- 1 root root   84624 Oct 31 15:34 adf_ctl
-rwxr-x--- 1 root root    4781 Oct 31 15:34 c3xxx_dev0.conf
-rwxr-x--- 1 root root    4265 Oct 31 15:34 c3xxxvf_dev0.conf.vm
-rwxr-x--- 1 root root    5145 Oct 31 15:34 c6xx_dev0.conf
-rwxr-x--- 1 root root    5151 Oct 31 15:34 c6xx_dev1.conf
-rwxr-x--- 1 root root    5153 Oct 31 15:34 c6xx_dev2.conf
-rwxr-x--- 1 root root    4265 Oct 31 15:34 c6xxvf_dev0.conf.vm
-rwxr-x--- 1 root root    5130 Oct 31 15:34 d15xx_dev0.conf
-rwxr-x--- 1 root root    3980 Oct 31 15:34 d15xxpf_dev0.conf
-rwxr-x--- 1 root root    4265 Oct 31 15:34 d15xxvf_dev0.conf.vm
-rwxr-x--- 1 root root    4350 Oct 31 15:34 dh895xcc_dev0.conf
-rwxr-x--- 1 root root    4349 Oct 31 15:34 dh895xcc_dev1.conf
-rwxr-x--- 1 root root    4266 Oct 31 15:34 dh895xccvf_dev0.conf.vm
-rwxr-x--- 1 root root  499752 Oct 31 15:34 intel_qat.ko
-rw-r--r-- 1 root root  107494 Oct 31 15:34 libadf.a
-rw-r--r-- 1 root root  115722 Oct 31 15:34 libosal.a
-rw-r--r-- 1 root root  833448 Oct 31 15:34 libqat.a
-rwxr-xr-x 1 root root  613696 Oct 31 15:34 libqat_s.so
-rw-r--r-- 1 root root   35658 Oct 31 15:34 libusdm_drv.a
-rwxr-xr-x 1 root root   26080 Oct 31 15:34 libusdm_drv_s.so
-rwxr-x--- 1 root root  807264 Oct 31 15:34 qat_895xcc.bin
-rwxr-x--- 1 root root  114176 Oct 31 15:34 qat_895xcc_mmp.bin
-rwxr-x--- 1 root root  531100 Oct 31 15:34 qat_c3xxx.bin
-rwxr-x--- 1 root root   29448 Oct 31 15:34 qat_c3xxx.ko
-rwxr-x--- 1 root root  120964 Oct 31 15:34 qat_c3xxx_mmp.bin
-rwxr-x--- 1 root root   20712 Oct 31 15:34 qat_c3xxxvf.ko
-rwxr-x--- 1 root root 1194696 Oct 31 15:34 qat_c62x.bin
-rwxr-x--- 1 root root   29696 Oct 31 15:34 qat_c62x.ko
-rwxr-x--- 1 root root  121604 Oct 31 15:34 qat_c62x_mmp.bin
-rwxr-x--- 1 root root   20704 Oct 31 15:34 qat_c62xvf.ko
-rwxr-x--- 1 root root 1194696 Oct 31 15:34 qat_d15xx.bin
-rwxr-x--- 1 root root   29056 Oct 31 15:34 qat_d15xx.ko
-rwxr-x--- 1 root root  120964 Oct 31 15:34 qat_d15xx_mmp.bin
-rwxr-x--- 1 root root   20712 Oct 31 15:34 qat_d15xxvf.ko
-rwxr-x--- 1 root root   28272 Oct 31 15:34 qat_dh895xcc.ko
-rwxr-x--- 1 root root   20280 Oct 31 15:34 qat_dh895xccvf.ko
-rwxr-x--- 1 root root    9228 Oct 31 15:34 qat_service
-rw-r--r-- 1 root root   38248 Oct 31 15:34 usdm_drv.ko
```

QAT works fine with other software but a little underloaded. I'd like to use QAT also with my 5TB zpool. Please add the support of newer QAT driver if possible.

**Comment by behlendorf** [comment]
Let's check with @wli5 what additional changes might be required to support  QAT 1.7

**Comment by wli5** [comment]
According to https://01.org/sites/default/files/downloads/intelr-quickassist-technology/336212004qatswgsg.pdf, please follow the new steps to install the QAT driver:
 $ ./configure --enable-kapi
 $make 
 $make install

**Comment by wli5** [comment]
It seems the QAT has not been started. Can you please verify the QAT service has been started?
If all goes well, you can get the QAT counts from:
cat /proc/spl/kstat/zfs/qat

I would suggest you build the ZFS from the latest source code instead from the dkms release, as there are several bug fixes not included in the current dkms package. Which version are you using?

**Comment by TheTDD** [comment]
Hello @wli5, it is not a problem with dkms because dkms deb/rpm package is built from latest source code. The version is actual 0.7.11. I got ZFS using QAT only after I starting ZFS after QAT manually. I saw the  /proc/spl/kstat/zfs/qat filled with come stats. The stats are not so good as I expected, but this is probably another issue. At the boot time when the ZFS initializes (loads zfs module) the QAT subsystem is not ready yet, therefore qat_init is always failed (qat_init_done=B_FALSE) and after that QAT compression is not used by ZFS (qat_use_accel returns B_FALSE). It is not possible to load zfs.ko after QAT because ZFS is in initramfs and used at boot time. QAT is started later with qat_service. 

Is it possible to implement some later initialization of qat_init to avoid this problem? Perhaps is is better to avoid the early initialization of QAT at all and use Data Plane API instead ($ICP_ROOT/quickassist/include/dc/cpa_dc_dp.h)?

Another problem I told is that there are too much failed requests to compression, and all requests to decompression are failed. I'd like to have 2 separate module parameters zfs_qat_disable_compression and zfs_qat_disable_decompression as workaround before the issue with decompression is solved. Also more detailed stats about failed requests will be nice (buffer overflow, timeout, retransmission etc).

**Comment by wli5** [comment]
@TheTDD these two bug fixes are not in dkms 0.7.11, so I believe this is the root cause why you see the overflow failures.
https://github.com/zfsonlinux/zfs/pull/7305 
https://github.com/zfsonlinux/zfs/pull/7338 

As for the later initialization of qat_init, I have not thought about it so far, as it is not a common use case as I know, but if you have a patch to make it working, I'm happy to see and review it. Thanks!

**Comment by TheTDD** [comment]
@wli5 thank you for your suggestion how to compile ZFS with QAT. I'm not actually satisfied with the implementation and will try to create new one from scratch using no pre-initialization and fast DC-DP routines with flat buffers. Decompression is already done and shows nice performance and no overruns at all. When also compression is done and tested well I'll contact you again and provide patch or merge request.

**Comment by wli5** [comment]
@TheTDD I'm happy to review your patch, but I suspect if the DP API will provide better performance than the current traditional API for the ZFS case, please also consider below facts:
1. Using separate polling thread in DP API.
2. Each 4K flat buffer will be in a stateless request, compression ratio will drop as history buffer can't maintain.

**Comment by TheTDD** [comment]
@wli5 Please look on my implementation in https://github.com/TheTDD/zfs/tree/qatDcDp
It is the branch made of zfs/0.7.11 tag.

1) The polling thread is not necessary because there is inline polling feature icp_sal_DcPollDpInstance available (Dc0IsPolled=1 in QAT-config file). However it is still possible to use interrupts (and it is a preferred way by my experience) by defining Dc0IsPolled=0.

2) There is no compression history in current implementation either. The instances and sessions are cached, but sessions are CPA_DC_STATELESS and all compressions are done with CPA_DC_FLUSH_FINAL. QAT is just compressing independent chunks one by one. That is exactly DP is doing but much faster because of flat buffers. That mean the compression ratio will remain the same. Also current implementation do not use the compression level (1-9) parameter in QAT-compression. I had implement it,

Here is status report after 3 hours of productive test with running roicoin daemon on gzip-9 compressed filesystem:

```
Every 5.0s: zfs_qat   Thu Nov  8 11:18:55 2018

18 1 0x01 25 6800 5401339925 2960252540632
name                            type data
init_failed                     4    35
comp_requests                   4    9637
comp_total_in_bytes             4    1253693952
comp_total_out_bytes            4    749135674
comp_fails                      4    488
decomp_requests                 4    13435
decomp_total_in_bytes           4    1086131712
decomp_total_out_bytes          4    1753540096
decomp_fails                    4    0
err_gen_header                  4    0
err_timeout                     4    0
err_gen_footer                  4    0
err_too_big_result              4    0
err_status_fail                 4    0
err_status_retry                4    0
err_status_param                4    0
err_status_resource             4    0
err_status_baddata              4    0
err_status_restarting           4    0
err_status_unknown              4    0
err_op_overflow                 4    488
err_op_hw                       4    0
err_op_sw                       4    0
err_op_fatal                    4    0
err_op_unknown                  4    0

QAT-Compression: ON
QAT-Decompression: ON
```
These stats are perfect - it shows some overflows on compress and no decompression errors. The "init_failed" happens in the time between loading of zfs module and starting roicoin and finishing initialization of QAT. The amount of allowed init failures is customizable via zfs module parameters.

```
[root@bg zfs]# modinfo zfs | grep qat
depends:        spl,znvpair,zcommon,zunicode,qat_api,zavl,icp
parm:           zfs_qat_disable_compression:Disable QAT compression (int)
parm:           zfs_qat_disable_decompression:Disable QAT decompression (int)
parm:           zfs_qat_init_failure_threshold:Threshold (number of init failures) to consider disabling QAT (int)
```
I saw later that you are about to use QAT for other tasks in ZFS 0.8.11, so if this DP implementation will succeeded I'll check the other stuff too.

**Comment by TheTDD** [comment]
@wli5 Thanks for reviewing. I would like to include this function into my production systems to get the best of QAT and ZFS. It will be nice if the code will be included in forthcoming release of ZoL. I'm continue working on it so please stay tuned. 

I have added some more stats to simplify benchmarking

```
Every 5.0s: zfs_qat  Fri Nov  9 09:15:16 2018

18 1 0x01 27 7344 5351126282 1169120209858
name                            type data
init_failed                     4    0
comp_requests                   4    1611
comp_total_in_bytes             4    208101888
comp_total_out_bytes            4    120831344
comp_fails                      4    0
decomp_requests                 4    3640
decomp_total_in_bytes           4    277953536
decomp_total_out_bytes          4    474187264
decomp_fails                    4    0
err_gen_header                  4    0
err_timeout                     4    0
err_gen_footer                  4    0
err_overflow                    4    0
err_status_fail                 4    0
err_status_retry                4    0
err_status_param                4    0
err_status_resource             4    0
err_status_baddata              4    0
err_status_restarting           4    0
err_status_unknown              4    0
err_op_overflow                 4    0
err_op_hw                       4    0
err_op_sw                       4    0
err_op_fatal                    4    0
err_op_unknown                  4    0
throughput_comp_bps             4    34552857
throughput_decomp_bps           4    398142119

QAT-Compression: ON
QAT-Decompression: ON
```
throughput_comp_bps  and throughput_decomp_bps are showing the throughput in bytes/sec. Actual compression is ~33MB/s, decompression is ~380MB/s. 
About compression ratio I have collected some stats on test system with average mixed text/archives data set:

Size of file systems
```
[root@bg tank]# zfs get written
NAME          PROPERTY  VALUE    SOURCE
tank          written   299M     -
tank/gzip-1   written   432M     -
tank/gzip-2   written   427M     -
tank/gzip-3   written   425M     -
tank/gzip-4   written   423M     -
tank/gzip-5   written   423M     -
tank/gzip-6   written   423M     -
tank/gzip-7   written   423M     -
tank/gzip-8   written   423M     -
tank/gzip-9   written   423M     -
tank/roicoin  written   574M     -
```
Compression settings
```
[root@bg slawa]# zfs get compression
NAME          PROPERTY     VALUE     SOURCE
tank          compression  off       local
tank/gzip-1   compression  gzip-1    local
tank/gzip-2   compression  gzip-2    local
tank/gzip-3   compression  gzip-3    local
tank/gzip-4   compression  gzip-4    local
tank/gzip-5   compression  gzip-5    local
tank/gzip-6   compression  gzip      local
tank/gzip-7   compression  gzip-7    local
tank/gzip-8   compression  gzip-8    local
tank/gzip-9   compression  gzip-9    local
tank/roicoin  compression  gzip-9    local
```
Compression ratio
```
[root@bg tank]# zfs get compressratio
NAME          PROPERTY       VALUE  SOURCE
tank          compressratio  1.40x  -
tank/gzip-1   compressratio  1.40x  -
tank/gzip-2   compressratio  1.41x  -
tank/gzip-3   compressratio  1.42x  -
tank/gzip-4   compressratio  1.42x  -
tank/gzip-5   compressratio  1.42x  -
tank/gzip-6   compressratio  1.42x  -
tank/gzip-7   compressratio  1.42x  -
tank/gzip-8   compressratio  1.43x  -
tank/gzip-9   compressratio  1.43x  -
tank/roicoin  compressratio  1.47x  -
```
tank/roicoin is the running instance of roicoin daemon with complete blockchain.

**Comment by cfzhu** [comment]
Hi @TheTDD , I try to run performance test with your code, The ZFS will print out some error message when I run the copy test:
<pre><code>root@qat-server-210:/home/czhu1x/fio_zfs_test# cp calgary /testpool/testfs/
root@qat-server-210:/home/czhu1x/fio_zfs_test# dmesg |more
[ 9698.659420] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659488] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659548] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659607] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659666] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659725] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section
[ 9698.659850] ZFS-QAT/dc: failed to find free DC instance ouf of 36, consider to increase NumberDcInstances in [KERNEL_QAT] section</code></pre>
There will be some data going to the software compression, this will have some impact on performance.

For benchmark, I run some tests using the FIO tool with your code. 
Compression settings
<pre><code>root@qat-server-210:/home/czhu1x/fio_zfs_test# zfs get compression
NAME             PROPERTY     VALUE     SOURCE
testpool         compression  gzip-1    local
testpool/testfs  compression  gzip-1    local</code></pre>
Compression  throughput:
<pre><code>WRITE: bw=2563MiB/s (2687MB/s), 2563MiB/s-2563MiB/s (2687MB/s-2687MB/s), io=256GiB (275GB), run=102292-102292msec</code></pre>
Compression ratio:
<pre><code>root@qat-server-210:/home/czhu1x/fio_zfs_test# zfs get compressratio
NAME             PROPERTY       VALUE  SOURCE
testpool         compressratio  1.97x  -
testpool/testfs  compressratio  1.97x  -</code></pre>

In the same environment and configuration, we can get a better performance with the current code. 
Compression throughput:
<pre><code>WRITE: bw=7402MiB/s (7762MB/s), 7402MiB/s-7402MiB/s (7762MB/s-7762MB/s), io=256GiB (275GB), run=35413-35413msec</code></pre>
Compression ratio:
<pre><code>root@qat-server-210:/home/czhu1x/fio_zfs_test# zfs get compressratio
NAME             PROPERTY       VALUE  SOURCE
testpool         compressratio  2.30x  -
testpool/testfs  compressratio  2.30x  -</code></pre>

**Comment by TheTDD** [comment]
Hi @cfzhu, thank you for testing and benchmarking my code. I'm continue working on it to use kmem_cache, reduce repeating tasks, cache everything that is possible and so on, but I'm afraid it will be not possible to overperform 3x advantage of current ZFS-QAT code because for DP operations the instance, intermediate buffers and session should be started/created and stopped/removed for every operation (thread), also each thread is only allowed to use unique instance (that is the reason of "failed to find free DC instance"). Current code is caching all instances, intermediate buffers and sessions and reuses it for every request. It is nice and fast, but unfortunately current code doesn't work properly in my configuration because ZFS is started much earlier before QAT is available and crashes if QAT get switched off (by example for changing number of instances). My code works properly in such cases, every operation is independent and goes to SW-compression if QAT is temporary not available. Anyway thank you very much for testing!

**Comment by wli5** [comment]
@cfzhu thanks for benchmarking!
@TheTDD based on the performance result, I'm afraid we can't merge your code.

**Comment by TheTDD** [comment]
@cfzhu @wli5 If you have time you may check the performance of new release https://github.com/TheTDD/zfs/tree/0.7.12-fast
It caches the instances and buffers (lazy initialization) and uses less spinlocks. It reaches stable 100MB/s compression and 430MB/s decompression as well as 370MB/s SHA2-256 checksums in long-running production test using 40 CY and 20 DC instances.

**Comment by geppi** [comment]
@wli5 Is there anything else that can be done to solve the initialization issue ?
I think it is not that uncommon to have the root filesystem on a mirrored rpool.
Faced with the choice to have either QAT for compression or rpool I would expect that there are a lot of candidates to opt for the second.

**Comment by geppi** [comment]
@wli5 @TheTDD I followed the suggestion in TheTDD zfs repo with no success. The QAT service is up and the parameter zfs_qat_disable gets set to zero by my zfs_qat service but there are still no qat statistics under /proc/spl/kstat/zfs. I'm using zfs-dkms 0.7.12-1~bpo9+1.
Do I need to install the reimplemented QAT support to make this work ?
I would like to stay with the ZoL mainline development but it looks like the early initialization of ZFS in the initramfs phase when there is no QAT service cannot be solved with this workaround.
Would it be possible to let the ZFS module retry the qat initialization when zfs_qat_disable gets changed from 1 to zero ?
Or do I need to get the setup for QAT support into the initramfs ? Is that even possible ?

**Comment by geppi** [comment]
@TheTDD Thank you for the clarification. However, I really would prefer to keep my installation standard as much as possible. Unfortunately it looks like your work is not going to get meerged into the mainstream.
@wli5 So it looks like there is currently no way to get QAT support working for ZFS with the mainstream implementation. I'm hesitating to start the endeavour to get the QAT support into the initramfs.
Couldn't the qat initialization in ZFS be retried at some point, e.g. when the zfs_qat_disable parameter is changed from 1 to zero ? This would make the method suggested by TheTDD also work for the mainstream implementation.
Currently having root on ZFS does imply no QAT support for ZFS. :-(

**Comment by geppi** [comment]
@wli5 I have opened a new issue for this feature request. See https://github.com/zfsonlinux/zfs/issues/8323

**Comment by TheTDD** [comment]
FYI: 0.7.12-fast successfully compiled and tested with QAT1.7.L.4.4.0-00023 drivers

**Comment by 8884361** [comment]
> FYI: 0.7.12-fast successfully compiled and tested with QAT1.7.L.4.4.0-00023 drivers

can you send me a dkms for test

**Comment by rlaager** [comment]
From the last comment, it sounds like this works now.
